satellite/metainfo/service.go

```diff
@@ -81,10 +88,23 @@ func (s *Service) UpdatePieces(ctx context.Context, path string, ref *pb.Pointer
 
 		// put all existing pieces to a map
 		pieceMap := make(map[int32]*pb.RemotePiece)
+		nodePieceMap := make(map[storj.NodeID]*pb.RemotePiece)
 		for _, piece := range pointer.GetRemote().GetRemotePieces() {
 			pieceMap[piece.PieceNum] = piece
+			if checkDuplicates {
+				nodePieceMap[piece.NodeId] = piece
+			}
 		}
 
+		// Return an error if the pointer already has a piece for this node
+		if checkDuplicates {
+			for _, piece := range toAdd {
+				_, ok := nodePieceMap[piece.NodeId]
+				if ok {
+					return nil, NodeAlreadyExitsError.New("node id already exists in piece. Path: %s, NodeID: %s", path, piece.NodeId.String())
+				}
+			}
+		}
 		// remove the toRemove pieces from the map
 		// only if all piece number, node id and hash match
 		for _, piece := range toRemove {
```

Comment:
Should we also update the `nodePieceMap` when update `pieceMap` in `toRemove` and `toAdd` loop?

Reply:
Hmmm. I don't think nodePieceMap gets used anymore at this point. Are you concerned someone might use in the future?

Reply:
Yea, but maybe that's not a concern for now